google/blockly
# Added Max Instances Property to Workspace Options

## The basics




## The details
### Resolves


[Issue 1435](https://github.com/google/blockly/issues/1435)

Also resolves a bug I could not find an issue for (I assume because no one noticed) where you could not duplicate a block, even if there was enough remaining capacity in the workspace to do so, because the context menu was checking against block.getDecendants() which includes blocks that do not get duplicated.

### Proposed Changes



- Adds a maxInstances map of block types -> max instances number to the workspace options object
- Adds an isDuplicatable() function to Blockly.Block.
- Disables the block in the toolbox if there are maxInstances number of that block in the workspace.
- Disables the Duplicate context menu option if the block is not duplicatable.
- Disables pasting the block from the clipboard if the block is not duplicatable.

### Reason for Changes


Adding this feature gives additional flexibility to event-driven and multi-threaded type Blockly applications. It also suggests creating Blockly games that force you to work more "efficiently" (with a limit number, and types, of blocks).

### Test Coverage



I tested the changes by following this proceedure:

1. Add a maxInstances property to the "controls_if" block with a value of 3.
2. Create 3 instances of the block, see if the block is disabled in the toolbox flyout. (Pass)
![disabledtoolbox](https://user-images.githubusercontent.com/25440652/48800902-e9c96a00-ecbf-11e8-8d57-6b5d53b3d264.jpg)
3. Create 1 instance of the block. Copy it to the clipboard. Press control-v repeatedly. See if the blocks stop pasting after 3 instances. (Pass)
![pastingworking](https://user-images.githubusercontent.com/25440652/48800923-f5b52c00-ecbf-11e8-928e-a5103b4f0913.jpg)
4. Create a "controls_if" block with another "controls_if" block inside it. See if the duplicate option is enabled for the inner if, but disabled for the outer if. (Pass)
![duplicateenabledinner](https://user-images.githubusercontent.com/25440652/48800935-ffd72a80-ecbf-11e8-89e5-881745172cbe.jpg)
![duplicatedisabledouter](https://user-images.githubusercontent.com/25440652/48800942-036ab180-ecc0-11e8-9803-e6a5d8e2439d.jpg)
5. Remove the maxInstaces property from the "controls_if" block. 
6. Add a maxInstances property to the "logic_boolean" block with a value of 1.
7. Create a "controls_if" block with another "controls_if" block inside it. Attatch a "logic_boolean" block to the inner if. See if the duplicate option is disabled for both blocks. (Pass)
![duplicatedisabledinnerboolean](https://user-images.githubusercontent.com/25440652/48800956-0b2a5600-ecc0-11e8-8a93-0a583faacc2b.jpg)
![duplicatedisabledouterboolean](https://user-images.githubusercontent.com/25440652/48800960-0cf41980-ecc0-11e8-9e66-50fede89aa7a.jpg)

This proceedure tests all of the features of this PR:
- Adds maxInstances property to JsonInit Object - It is being read properly in all steps.
- Adds an isDuplicatable() function to Blockly.Block. - This function is working correctly in all applicable steps (3, 4, 7)
- Disables the block in the toolbox if there are maxInstances number of that block in the workspace. - This is working correctly in all applicable steps (1)
- Disables the Duplicate context menu option if the block is not duplicatable. - This is working correctly in all applicable steps (4, 7)
- Disables pasting the block from the clipboard if the block is not duplicatable. - This is working correctly in all applicable steps (7)

Tested on:
* Desktop Chrome









### Additional Information
